src/extra/components/adq.py

```diff
@@ -0,0 +1,1294 @@
+                f'board{self.board}.interleavedMode'))
+
+        self._interleaved = interleaved
+
+        if clock_ratio is None:
+            if device_id in self._3g_digitizer:
+                clock_ratio = 392
+            else:
+                clock_ratio = 440
+
+        self._clock_ratio = clock_ratio * (2 if interleaved else 1)
+
+        self._first_pulse_offset = first_pulse_offset
+        self._single_pulse_length = single_pulse_length
+
+        if pulses is False:
+            pulses = None  # Explicitly disabled.
+        elif pulses is None:
+            # If none, try to auto-detect XrayPulses for this data.
+            try:
+                pulses = XrayPulses(data)
+            except ValueError:
+                raise ValueError('could not auto-detect pulse information, '
+                                 'please pass explicit pulses object or '
+                                 'disable explicitly with pulses=False')
+
+        self._pulses = pulses
+
+        self._preprocess_kw = self._validate_preprocess(preprocess_kw)
+
+    def __repr__(self):
+        source = self._instrument_src.source
+        return "<{} {} {}>".format(
+            type(self).__name__, source[:source.find(':')],
+            self._channel_name)
+
+    @classmethod
+    def _find_adq_pipeline(cls, data, prefix=''):
+        """Try to find ADQ digitizer instrument source name."""
+
+        digitizers = set()
+
+        for source in data.instrument_sources:
+            m = cls._adq_pipeline_re.match(source)
+            if m is not None and source.startswith(prefix):
+                digitizers.add(m[0])
+
+        if not digitizers:
+            raise ValueError(f'no digitizer source found for prefix={prefix} '
+                             f'and pattern={cls._adq_pipeline_re.pattern}, '
+                             f'please pass explicit instrument source')
+        elif len(digitizers) > 1:
+            raise ValueError('multiple digitzer sources found for '
+                             'prefix={} and pattern={}:\n{}'.format(
+                                prefix, cls._adq_pipeline_re.pattern,
+                                ', '.join(sorted(digitizers))))
+
+        return digitizers.pop()
+
+    @staticmethod
+    def _correct_cm_by_train(signal, out, period, baseline, baselevel=None):
+        """Correct common mode in signal by each train trace."""
+
+        if isinstance(baseline, slice):
+            baseline = signal[..., baseline]
+
+        if baselevel is not None:
+            baseline = baseline - baselevel
+
+        for offset in range(period):
+            sel = np.s_[offset::period]
+            np.subtract(
+                signal[..., sel],
+                baseline[..., sel].mean(axis=signal.ndim - 1)[..., None],
+                out=out[..., sel], casting='safe')
+
+    @staticmethod
+    def _correct_cm_by_mean(signal, out, period, baseline, baselevel=None):
```

Comment:
This method is currently not used, but could be exposed in the future. Computing the common mode over multiple trains rather than per train can have advantages in certain situations, but generally I typically only use the train itself (e.g. also in offline calibration for DLDs).